python/ray/autoscaler/_private/node_provider_availability_tracker.py

```diff
@@ -0,0 +1,201 @@
+                unavailable_node_information=unavaiable_information,
+                **node_availability_record_dict,
+            )
+
+        return NodeAvailabilitySummary(node_availabilities=parsed)
+
+    def summary_string(self, separator_len: int) -> str:
+        if self:
+            formatted_lines = []
+            sorted_keys = sorted(
+                self.node_availabilities.keys(),
+                key=lambda node_type: self.node_availabilities[
+                    node_type
+                ].last_checked_timestamp,
+            )
+            for node_type in sorted_keys:
+                record = self.node_availabilities[node_type]
+                category = "Available"
+                if not record.is_available:
+                    assert record.unavailable_node_information is not None
+                    category = record.unavailable_node_information.category
+                attempted_time = datetime.datetime.fromtimestamp(
+                    record.last_checked_timestamp
+                )
+                formatted_time = (
+                    f"{attempted_time.hour}:"
+                    f"{attempted_time.minute}:"
+                    f"{attempted_time.second}"
+                )
+                formatted_line = (
+                    f" {node_type} (attempted={formatted_time}): {category}"
+                )
+                formatted_lines.append(formatted_line)
+
+            return (
+                "Launches\n"
+                + ("-" * separator_len)
+                + "\n"
+                + "Node types:\n"
+                + ("\n".join(formatted_lines))
+            )
+
+        else:
+            return ""
+        pass
+
+    def __eq__(self, other: "NodeAvailabilitySummary"):
+        return self.node_availabilities == other.node_availabilities
+
+    def __bool__(self) -> bool:
+        return bool(self.node_availabilities)
+
+
+class NodeProviderAvailabilityTracker:
+    """A thread safe, TTL cache of node provider availability. We don't use
+    cachetools.TTLCache because we want fine grain control over when entries
+    expire (e.g. insert an entry at a previous point in time)."""
+
+    def __init__(
+        self,
+        timer: Callable[[], float] = time.time,
+        ttl: float = AUTOSCALER_NODE_AVAILABILITY_MAX_STALENESS_S,
+    ):
+        """A cache that tracks the availability of nodes and throw away
+        entries which have grown too stale.
+
+        Args:
+          timer: A function that returns the current time in seconds.
+          ttl: The ttl from the insertion timestamp of an entry.
+        """
+        self.timer = timer
+        self.ttl = ttl
+        # Mapping from node type to (eviction_time, record)
+        self.store: Dict[str, Tuple[float, NodeAvailabilityRecord]] = {}
+        # A global lock to simplify thread safety handling.
+        self.lock = threading.RLock()
+
+    def _unsafe_update_node_availability(
```

Comment:
`_unsafe_` frightens me.
Maybe `_inner_` or just `_`?
Up to you, though.

Reply:
how about suffix `_requires_lock`. just trying to make it clear that this is not a thread safe function.

Reply:
okidoke